### DISCUSSION TITLE
NextJS 9.4 and strict CSP policies

### DISCUSSION THREAD

**Discussion by montezume** [vercel/next.js#13248] [Help] [answered]
Is it possible to use the latest version of NextJS in an environment with strict CSP policies? At my workplace, we are reluctant to use different CSP policies in development than in production. This means that the eval's that fast refresh relies on don't work, so we are unable to upgrade from 9.3 to 9.4 

Any ideas? Is it possible to work with the latest NextJS without allowing unsafe-eval in your CSPs?

**Comment by levelasquez** [comment]
I'm having an issue relate to this.

i'm using next-secure-headers npm packages

this is my security config:

```javascript
export default {
  contentSecurityPolicy: {
    directives: {
      defaultSrc: "'self'",
      styleSrc: [
        "'self'",
        'https://stackpath.bootstrapcdn.com',
        "'sha256-4/2nIlfwIVTJ1+JcNQ6LkeVWzNS148LKAJeL5yofdN4='",
      ],
      imgSrc: [
        "'self'",
        'https://stats.g.doubleclick.net',
        'https://www.google.com',
        'https://www.google.cl',
        'https://www.facebook.com',
      ],
      fontSrc: [
        "'self'",
        'https://fonts.gstatic.com',
      ],
      scriptSrc: [
        "'self'",
        'https://www.google.com',
        'https://www.gstatic.com',
        'https://ajax.cloudflare.com',
        'https://www.googletagmanager.com',
        'https://www.google-analytics.com',
        'https://connect.facebook.net',
      ],
      frameSrc: ["'self'", 'https://www.google.com'],
      connectSrc: [
        "'self'",
        'https://www.gstatic.com',
      ],
    },
  },
  referrerPolicy: 'same-origin',
};
```

in v9.3 it works but when i upgrade to 9.4 it throw these errors:

```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' https://www.google.com https://www.gstatic.com https://ajax.cloudflare.com https://www.googletagmanager.com https://www.google-analytics.com https://connect.facebook.net".

    at Object../node_modules/@next/react-refresh-utils/runtime.js (react-refresh.js?ts=1590178953109:23)
    at __webpack_require__ (webpack.js?ts=1590178953109:863)
    at checkDeferredModules (webpack.js?ts=1590178953109:46)
    at webpackJsonpCallback (webpack.js?ts=1590178953109:33)
    at webpack.js?ts=1590178953109:1005
    at webpack.js?ts=1590178953109:1013
```

```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' https://www.google.com https://www.gstatic.com https://ajax.cloudflare.com https://www.googletagmanager.com https://www.google-analytics.com https://connect.facebook.net".

    at Object../node_modules/next/dist/client/next-dev.js (main.js?ts=1590178953109:2062)
    at __webpack_require__ (webpack.js?ts=1590178953109:863)
    at checkDeferredModules (webpack.js?ts=1590178953109:46)
    at Array.webpackJsonpCallback [as push] (webpack.js?ts=1590178953109:33)
    at main.js?ts=1590178953109:1
```

any idea?

**Reply by FernandoBoza** [reply]
any updated on this. same issue

**Comment by montezume** [comment, marked as answer]
Sorry, I realized that my problem is not with fast-refresh, but with the webpack devtool. I changed to a non-eval based devtool and it worked

**Reply by fabien-h** [reply]
How did you do that ? :)

Because when I try to change it, I get this in my terminal : `Warning: Reverting webpack devtool to 'eval-source-map'.`

Related to this : https://github.com/vercel/next.js/blob/master/errors/improper-devtool.md

**Reply by imaksp** [reply]
> How did you do that ? :)
> 
> Because when I try to change it, I get this in my terminal : `Warning: Reverting webpack devtool to 'eval-source-map'.`
> 
> Related to this : https://github.com/vercel/next.js/blob/master/errors/improper-devtool.md

Hi, better option would be to set empty sec headers if NODE_ENV is development.

**Reply by montezume** [reply]
I disagree, it's always best to have your dev environment mirror your production environment as much as possible.

**Reply by imaksp** [reply]
@montezume Ok but its a trade-off eval based devtool has highest performance in development 
See: https://webpack.js.org/configuration/devtool/
